File: frontend/cypress/test-solutions/minimum-window-substring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        if not s or not t:
            return ""
        
        # Count of each character in t
        t_count = {}
        for char in t:
            t_count[char] = t_count.get(char, 0) + 1
        
        required = len(t_count)
        formed = 0
        window_counts = {}
        
        left = right = 0
        min_len = float('inf')
        min_left = 0
        
        while right < len(s):
            # Add the right character to the window
            char = s[right]
            window_counts[char] = window_counts.get(char, 0) + 1
            
            # Check if this character contributes to the desired frequency
            if char in t_count and window_counts[char] == t_count[char]:
                formed += 1
            
            # Try to shrink the window from left
            while left <= right and formed == required:
                char = s[left]
                
                # Update the result if this window is smaller
                if right - left + 1 < min_len:
                    min_len = right - left + 1
                    min_left = left
                
                # Remove the leftmost character from the window
                window_counts[char] -= 1
                if char in t_count and window_counts[char] < t_count[char]:
                    formed -= 1
                
                left += 1
            
            right += 1
        
        return "" if min_len == float('inf') else s[min_left:min_left + min_len]

File: backend/test-solutions/minimum-window-substring.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        if (s.empty() || t.empty()) return "";
        
        // Count of each character in t
        unordered_map<char, int> tCount;
        for (char c : t) {
            tCount[c]++;
        }
        
        int required = tCount.size();
        int formed = 0;
        unordered_map<char, int> windowCounts;
        
        int left = 0, right = 0;
        int minLen = INT_MAX;
        int minLeft = 0;
        
        while (right < s.length()) {
            // Add the right character to the window
            char c = s[right];
            windowCounts[c]++;
            
            // Check if this character contributes to the desired frequency
            if (tCount.find(c) != tCount.end() && windowCounts[c] == tCount[c]) {
                formed++;
            }
            
            // Try to shrink the window from left
            while (left <= right && formed == required) {
                char leftChar = s[left];
                
                // Update the result if this window is smaller
                if (right - left + 1 < minLen) {
                    minLen = right - left + 1;
                    minLeft = left;
                }
                
                // Remove the leftmost character from the window
                windowCounts[leftChar]--;
                if (tCount.find(leftChar) != tCount.end() && windowCounts[leftChar] < tCount[leftChar]) {
                    formed--;
                }
                
                left++;
            }
            
            right++;
        }
        
        return minLen == INT_MAX ? "" : s.substr(minLeft, minLen);
    }
};